noir_stdlib/src/collections/map.nr

```diff
@@ -0,0 +1,305 @@
+            if slot.is_valid(){
+                let (_, value) = slot.key_value_unchecked();
+                values.push(value);
+            }
+        }
+        values
+    }
+
+    // Retains only the elements specified by the predicate.
+    pub fn retain(&mut self, f: fn(K, V) -> bool){
+        for index in 0..N{
+            let mut slot = self._table[index];
+            if slot.is_valid(){
+                let (key, value) = slot.key_value_unchecked();
+                if !f(key, value){
+                    slot.mark_deleted();
+                    self._len -= 1;
+                    self._table[index] = slot;
+                }
+            }
+        }
+    }
+
+    fn hash(self, key: K) -> u64
+    where
+        K: Hash,
+        B: BuildHasher<H>,
+        H: Hasher
+    {
+        let mut hasher = self._build_hasher.build_hasher();
+        key.hash(&mut hasher);
+        hasher.finish() as u64
+    }
+
+    // Probing scheme: quadratic function.
+    // We use 0.5 constant near variadic attempt and attempt^2 monomials.
+    // This ensures good uniformity of distribution for table sizes
+    // equal to prime numbers or powers of two. 
+    fn quadratic_probe(self, hash: u64, attempt: u64) -> u64 {
+        (hash + (attempt + attempt * attempt) / 2 ) % N
+    }
+}
+
+// While conducting lookup, we iterate attempt from 0 to N - 1 due to heuristic,
+// that if we have went that far without finding desired, 
+// it is very unlikely to be after - performance in that case would be drastical.
+impl<K, V, N, B, H> Map<K, V> for HashMap<K, V, N, B>
+where
+    K: Eq + Hash,
+    B: BuildHasher<H>,
+    H: Hasher
+{
+    fn len(self) -> u64{
+        self._len
+    }
+
+    fn get(self, key: K) -> Option<V>{
+        let mut result = Option::none();
+
+        let hash = self.hash(key);
+        let mut break = false;
+
+        for attempt in 0..N{
+            if !break{
+                let index = self.quadratic_probe(hash, attempt as u64);
+                let slot = self._table[index];
+
+                // Not marked as deleted and has key-value.
+                if slot.is_valid(){ 
+                    let (current_key, value) = slot.key_value_unchecked();
+                    if current_key == key{
+                        result = Option::some(value);
+                        break = true;
+                    }
+                }
+            }
+        }
+
+        result
+    }
+
+    fn insert(&mut self, key: K, value: V){
+        let hash = self.hash(key);
+        let mut break = false;
+
+        for attempt in 0..N{
+            if !break{
+                let index = self.quadratic_probe(hash, attempt as u64);
+                let mut slot = self._table[index];
+                let mut insert = false;
+
+                // Either marked as deleted or has unset key-value.
+                if slot.is_available(){ 
+                    insert = true;
+                    self._len += 1;
```

Comment:
After thinking this over some more, I think this is missing an assert that the hash map isn't too filled already. E.g. when I've written linear probing hashmaps before I'd usually reallocate when the map gets half-full. We'll certainly want an assert sometime before this hashmap gets completely full I think. Otherwise performance will degrade and it will be more difficult to debug when nothing more can be inserted into the map.